In [ ]:
!pip install transformers datasets accelerate scikit-learn

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 62.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 49.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 27.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 18.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 6.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 86.9 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstalling nvidia-nvjitlink-cu12-12.5.82:
      Successfully uninstalled nvidia-nvjitlin

In [ ]:
import pandas as pd
import torch
from torch.utils.data import Dataset, DataLoader
from transformers import AutoTokenizer, AutoModelForSequenceClassification
from torch.optim import AdamW
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score, classification_report
from sklearn.model_selection import train_test_split
import torch.nn as nn

In [ ]:
train_df = pd.read_csv("drive/MyDrive/train_dataset_train.csv")
test_df = pd.read_csv("drive/MyDrive/test_dataset_test.csv")

In [ ]:
train_df.head()

id                                    Текст Сообщения  \
0  2246  Помогите начальник Льговского рэс не реагирует...   
1   380  <p>По фасаду дома по адресу ул. Урицкого 22 пр...   
2  2240  Агресивные собаки. На радуге там стая из подро...   
3   596  <p>На пересечении &nbsp;улиц Сосновская и Бере...   
4  1797  <p style=`text-align:justify;`><span style=`ba...   

                                            Тематика  \
0  Нарушения, связанные с содержанием электросети...   
1                                  Аварийные деревья   
2                              Безнадзорные животные   
3  Нескошенная сорная растительность в местах общ...   
4                                  Аварийные деревья   

                           Ответственное лицо  Категория  
0             Администрация Льговского района          3  
1                 Администрация города Курска          3  
2                 Администрация города Курска          1  
3   Комитет дорожного хозяйства города Курска          3  
4  Комитет городского хозяйства города Курска          3

In [ ]:
train_df = train_df.rename(columns={
    'Текст Сообщения': 'text',
    'Категория': 'label'
})

In [ ]:
# Преобразуем текстовые метки в числовые
label_encoder = LabelEncoder()
train_df['label'] = label_encoder.fit_transform(train_df['label'])

In [ ]:
# Разбиение на train и val
X_train, X_val, y_train, y_val = train_test_split(train_df['text'], train_df['label'], test_size=0.2, random_state=42)


In [ ]:
# Токенизатор и модель
model_name = "DeepPavlov/rubert-base-cased"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=len(label_encoder.classes_))


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at DeepPavlov/rubert-base-cased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [ ]:
# Кастомный Dataset
class TextDataset(Dataset):
    def __init__(self, texts, labels, tokenizer, max_length=256):
        self.texts = texts
        self.labels = labels
        self.tokenizer = tokenizer
        self.max_length = max_length
    def __len__(self):
        return len(self.texts)
    def __getitem__(self, idx):
        encoding = self.tokenizer(self.texts[idx], truncation=True, padding='max_length', max_length=self.max_length, return_tensors='pt')
        return {
            "input_ids": encoding["input_ids"].squeeze(0),
            "attention_mask": encoding["attention_mask"].squeeze(0),
            "labels": torch.tensor(self.labels[idx], dtype=torch.long)
        }

# Создание датасетов и загрузчиков
train_dataset = TextDataset(X_train.tolist(), y_train.tolist(), tokenizer)
val_dataset = TextDataset(X_val.tolist(), y_val.tolist(), tokenizer)

train_loader = DataLoader(train_dataset, batch_size=16, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=16, shuffle=False)

# Определяем устройство (GPU, MPS, CPU)
device = torch.device("cuda" if torch.cuda.is_available() else "mps" if torch.backends.mps.is_available() else "cpu")
model.to(device)

# Оптимизатор и loss
optimizer = AdamW(model.parameters(), lr=2e-5)
criterion = nn.CrossEntropyLoss()

# Функция обучения
def train(model, dataloader, optimizer, criterion):
    model.train()
    total_loss = 0
    for batch in dataloader:
        input_ids = batch["input_ids"].to(device)
        attention_mask = batch["attention_mask"].to(device)
        labels = batch["labels"].to(device)

        optimizer.zero_grad()
        outputs = model(input_ids, attention_mask=attention_mask)
        loss = criterion(outputs.logits, labels)
        loss.backward()
        optimizer.step()

        total_loss += loss.item()
    return total_loss / len(dataloader)

In [ ]:
# Функция оценки
def evaluate(model, dataloader):
    model.eval()
    predictions, true_labels = [], []
    with torch.no_grad():
        for batch in dataloader:
            input_ids = batch["input_ids"].to(device)
            attention_mask = batch["attention_mask"].to(device)
            labels = batch["labels"].to(device)

            outputs = model(input_ids, attention_mask=attention_mask)
            preds = torch.argmax(outputs.logits, dim=1)

            predictions.extend(preds.cpu().numpy())
            true_labels.extend(labels.cpu().numpy())

    acc = accuracy_score(true_labels, predictions)
    unique_labels = sorted(set(true_labels))
    target_names = [str(label_encoder.classes_[i]) for i in unique_labels]
    report = classification_report(
        true_labels,
        predictions,
        labels=unique_labels,
        target_names=target_names
    )
    return acc, report


In [ ]:
# Цикл обучения
num_epochs = 3
for epoch in range(num_epochs):
    train_loss = train(model, train_loader, optimizer, criterion)
    val_acc, val_report = evaluate(model, val_loader)
    print(f"Epoch {epoch+1}/{num_epochs} - Train Loss: {train_loss:.4f} - Val Accuracy: {val_acc:.4f}")

print("Validation classification report:\n", val_report)

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Epoch 1/3 - Train Loss: 0.7325 - Val Accuracy: 0.8250


/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Epoch 2/3 - Train Loss: 0.4608 - Val Accuracy: 0.8250
Epoch 3/3 - Train Loss: 0.3151 - Val Accuracy: 0.8450
Validation classification report:
               precision    recall  f1-score   support

           0       0.82      0.84      0.83        99
           1       0.86      1.00      0.92         6
           2       0.00      0.00      0.00         2
           3       0.87      0.88      0.87       183
           4       0.96      0.92      0.94        24
           5       0.00      0.00      0.00         3
           7       0.40      0.67      0.50         3
           8       0.89      0.97      0.93        32
           9       0.00      0.00      0.00         1
          10       1.00      0.83      0.91        12
          11       0.00      0.00      0.00         4
          13       0.00      0.00      0.00         2
          15       1.00      0.50      0.67         2
          16       0.71      0.81      0.76        27

   micro avg       0.85      0.84      0.85  

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


**Вывод:**
Вывод по результатам обучения и валидации модели
В ходе выполнения эксперимента была обучена модель на основе предобученного трансформера DeepPavlov/rubert-base-cased для задачи классификации текстовых сообщений по 17 категориям. Обучение проводилось с использованием PyTorch, с разбиением данных на обучающую и валидационную выборки в соотношении 80:20.

Основные результаты:

Модель достигла точности (accuracy) на валидационном наборе около 82–85% за 3 эпохи обучения, что свидетельствует о достаточно хорошем качестве классификации для многоклассовой задачи.
Однако, для некоторых классов, особенно с малым количеством примеров в выборке, наблюдаются случаи, когда модель не смогла корректно их распознать. Это выразилось в предупреждениях UndefinedMetricWarning и нулевых значениях метрик precision и recall для ряда классов.
Дисбаланс классов, характерный для исходного датасета (отсутствие некоторых классов в валидационной выборке), влияет на качество предсказаний для редких категорий.
Метрики macro-average (усреднённые по классам) показывают пониженные значения по сравнению с weighted-average, что указывает на снижение качества для менее представленных классов.

Рекомендации для улучшения качества модели:

Собрать и добавить дополнительные данные для редких классов с целью выравнивания представительства.
Рассмотреть использование методов балансировки выборки (oversampling, class weights).
Провести дополнительную настройку гиперпараметров и исследовать более глубокие или специализированные модели.
Анализировать ошибки модели для выявления специфических сложностей в распознавании отдельных категорий.
В целом, полученные результаты демонстрируют успешную базовую реализацию задачи классификации текстов на русском языке с помощью современных трансформерных моделей. При дальнейшем улучшении данных и оптимизации подходов можно добиться более стабильной и качественной работы модели по всем классам.